aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -53,17 +53,26 @@ static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {
   return out_impl(result, self);
 }
 
+// Implicitly promote all int and bool dtypes to default floating dtype
+// See PR #33322 for more discussion on this
+template <typename OutImpl>
+static inline Tensor unary_float_op_impl(const Tensor& self, OutImpl& out_impl) {
+  Tensor result = isIntegralType(self.scalar_type(), /*include_bool=*/ true) ? at::empty({0}, self.options().dtype(typeMetaToScalarType(c10::get_default_dtype()))) : \
+                  at::empty({0}, self.options());
+  return out_impl(result, self.to(result.scalar_type()));
```

Comment:
Is this cast of self necessary? What happens without it?

Reply:
I think it's necessary: `torch.expm1(torch.tensor(2))` throws `RuntimeError: expected dtype Float but got dtype Long` error.

Reply:
It's probably because of this check here: https://github.com/pytorch/pytorch/blame/master/aten/src/ATen/native/TensorIterator.cpp#L143

Reply:
I think it will create a TensorIterator::unary_op with mismatched types, and because TensorIterator::unary_op doesn't yet support internal promotion, that will error.  That's part of what we have to think about for the next stage (promotion based on an explicit dtype argument).

Reply:
lol, i was writing my comment on a stale version of the webpage...

Reply:
This should be resolved here. https://github.com/pytorch/pytorch/pull/33322/commits/c2ff01ba86bc13ce422c8fb49126aabf547b2b29